File: trigger-model-development/flood/trigger-model/scripts/catchment_extractor.R
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(tmap)
library(maps)
library(httr)
library(sf)
library(lubridate)
library(zoo)
library(xts)
library(lubridate)
require(data.table)

##############################################
clip <- function(raster, shape) {
  raster_crop <- crop(raster,shape)
  raster_bsn <- mask(raster_crop,shape)
  return(raster_bsn)
}

##############################################

# extract catchment for an areae

catchment_extractor<-function(sample_admin,basin){
  #basin <- st_read(dsn='C:/documents/General_data/Basins/hydrosheds/African_basins',layer='hybas_lake_af_lev12_v1c')
  basin_check2 <- st_intersection(sample_admin, basin)# %>% arrange(desc(UP_AREA)) %>% dplyr::select(HYBAS_ID,geometry)
  pts = 
    
    basin_check <- st_touches(sample_admin, basin)# %>% arrange(desc(UP_AREA)) %>% dplyr::select(HYBAS_ID,geometry)
  basin_<-basin %>% dplyr::select(HYBAS_ID,geometry)
  datalist = list()
  
  for (j in 1:length(as.list( basin_check2$HYBAS_ID)))  { 
    i<- as.list( basin_check2$HYBAS_ID)[[j]]
    list_con<-c(i)
    item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID 
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    checker<-c()
    ii=1
    if (length(item) !=0){
      while (TRUE)
      {
        ii=ii+1
        if (length(item) !=0){
          for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID   #1
          list_con<-c( list_con , item[ ! item %in% list_con ] )}
          item <-list_con
          checker[[ii]]<-length(list_con)
          if ((length(checker)>4) &(checker[[ii]]==checker[[ii-1]]))
          {
            break
          }
        }
        
        
      }
      
    }
    
    #assign(paste("cat", as.character(j), sep = "_"), list_con)
    cachment<-assign(paste("cat", as.character(j), sep = "_"), basin_[basin_$HYBAS_ID %in% list_con,])
    
    dat <- data.frame(HYBAS_ID = j, geom = cachment$geometry)
    df = st_as_sf(dat)
    datalist[[j]] <- df # add it to your list
    
    
  }
  return(datalist)
  
}


#---------------------- Load admin boundary data -------------------------------  

eth_admin3 <- st_read(dsn='C:/documents/ethiopia/admin3',layer='admin3')
# for visual add the river network in the plot 
#rivers <- st_read(dsn='C:/documents/General_data/Basins/hydrosheds/African_rivers',layer='af_riv_15s')
eth_admin3<-eth_admin3 %>%  dplyr::mutate(Pcode=NewPCODE) %>%  dplyr::select(Pcode,geometry)
basin <- st_read(dsn='C:/documents/General_data/Basins/hydrosheds/African_basins',layer='hybas_lake_af_lev12_v1c')


################################# run the code for sample admin 
sample_admin<-eth_admin3[1,]
cat_admin1<-catchment_extractor(sample_admin,basin)
item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID

File: trigger-model-development/flood/trigger-model/scripts/catchment_extractor.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 20 10:54:39 2019

@author: ATeklesadik
"""
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 20 10:54:39 2019

@author: ATeklesadik
"""
import geopandas as gpd
import fiona
import descartes
from osgeo import ogr
from shapely.geometry import Polygon
import numpy as np

################################# 
def add_to_list(items,list_):
    for elments in items:
        if elments not in list_:
            list_.append(elments)
    return list_

     

def catchment_extractor(sample_admin):
    datalist={}
    basin = gpd.read_file('C:/documents/General_data/Basins/hydrosheds/African_basins/hybas_lake_af_lev12_v1c.shp')
    basin_check2 = gpd.overlay(basin,sample_admin, how='identity')
    basin_check = basin_check2[basin_check2['Pcode'].isin(sample_admin['Pcode'].values)]
    basin_check = basin_check.sort_values(by=['UP_AREA'],ascending=False)
    basin_check.drop_duplicates(subset ="HYBAS_ID", keep = 'first', inplace = True) 
    basin_check1 = basin_check.iloc[:1]
    print(basin_check1)
    for j in basin_check1['HYBAS_ID'].values:
        list_con=[]
        list_con.append(j)
        item = np.unique(basin[basin['NEXT_DOWN'] == j]['HYBAS_ID'].values)
        list_con=add_to_list(item,list_con)
        ii=0
        checker={}
        if len(item) >0:
            while True: 
                ii=ii+1
                if len(item) >0:
                    for i in item:
                        itm2 = basin[basin['NEXT_DOWN'] == i]['HYBAS_ID'].values
                        list_con=add_to_list(itm2,list_con)
                    item=list_con 
                    checker[ii]=len(list_con)
                    if (len(checker) >4) and (checker[ii]==checker[ii-1]):
                        break
        cachment= basin[basin['HYBAS_ID'].isin(list_con)] 
        datalist[j] = cachment            
    return(datalist)
    
################################# run the code for sample admin 

eth_admin3 = gpd.read_file('C:/documents/ethiopia/admin3/admin3.shp')
rivers =gpd.read_file('C:/documents/General_data/Basins/hydrosheds/African_rivers/af_riv_15s.shp')

rivers = gpd.overlay(eth_admin3, rivers, how='intersection')
sample_admin = eth_admin3[0:1]
           
event_tc=catchment_extractor(sample_admin)

################# plot to see the result 
for key,valu in event_tc.items():
    print(key)
    valu.plot()
